### PR TITLE
apf: change controller to use SSA for patches

### LINKED ISSUE #107727: absence of openapi configuration in integration tests makes server-side apply panic (broke apf controller when it switched to SSA)
### What happened?

Integration test timed out due to the following error 
```
E0120 14:05:02.398256  118058 runtime.go:77] Observed a panic: FieldManager must be installed to run apply
```
The error appeared 748 times in the log. Looks like lots of churning happens during the apf configuration bootstrapping phase. 
Is there a race condition between when the type is being registered and the API calls? 

Stack trace:
```
I0120 14:05:02.397911  118058 apf_controller.go:879] Triggered API priority and fairness config reloading because priority level exempt is undesired and idle
I0120 14:05:02.398135  118058 panic.go:1038] "HTTP" verb="APPLY" URI="/apis/flowcontrol.apiserver.k8s.io/v1beta2/flowschemas/system-node-high/status?fieldManager=api-priority-and-fairness-config-consumer-v1&force=true" latency="650.951µs" userAgent="Go-http-client/1.1" audit-ID="bc3c561e-1a80-403e-b611-63ba7bd484b9" srcIP="127.0.0.1:46572" apf_pl="exempt" apf_fs="exempt" apf_fd="" resp=0
E0120 14:05:02.398256  118058 runtime.go:77] Observed a panic: FieldManager must be installed to run apply
goroutine 136765 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher.finishRequest.func1.1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:105 +0xaf
panic({0x3fc0bc0, 0x542eb00})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*applyPatcher).applyPatchToCurrentObject(0x551ceb0, {0x553da98, 0xc0887ccc00}, {0x551ceb0, 0xc0887da480})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:482 +0x449
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*patcher).applyPatch(0xc0bf781a40, {0x553da98, 0xc0887ccc00}, {0x4c296a0, 0xc0887da480}, {0x551ceb0, 0xc0887da480})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:566 +0xd3
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest.(*defaultUpdatedObjectInfo).UpdatedObject(0x81f4878, {0x553da98, 0xc0887ccc00}, {0x551ceb0, 0xc0887da480})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/rest/update.go:229 +0xd0
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Update.func1({0x551ceb0, 0xc0887da480}, {0xc07278e397b5dfde, 0x8b5f27eafb})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:533 +0x1f9
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*store).updateState(0xc03d204bd0, 0xc06ba15310, 0x42)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:894 +0x3e
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3.(*store).GuaranteedUpdate(0xc03d204bd0, {0x553da98, 0xc0887ccc00}, {0xc01908f2e0, 0x1d}, {0x551ceb0, 0xc0887da300}, 0x1, 0x40ef94, 0xc0848d34a0, ...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3/store.go:365 +0x56e
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.(*Cacher).GuaranteedUpdate(0xc0e9754c60, {0x553da98, 0xc0887ccc00}, {0xc01908f2e0, 0x1d}, {0x551ceb0, 0xc0887da300}, 0xa8, 0x4b86ae0, 0xc0848d34a0, ...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go:721 +0x1b5
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*DryRunnableStorage).GuaranteedUpdate(0x4cbc218, {0x553da98, 0xc0887ccc00}, {0xc01908f2e0, 0x4cac0b0}, {0x551ceb0, 0xc0887da300}, 0xf2, 0x1, 0xc0848d34a0, ...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/dryrun.go:97 +0x1c7
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry.(*Store).Update(0xc0e5c5b2c0, {0x553da98, 0xc0887ccc00}, {0xc0412d2b3d, 0xb}, {0x5523328, 0xc0887cccf0}, 0xc0105369b0, 0x4ea4ac8, 0x0, ...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/registry/generic/registry/store.go:521 +0x508
k8s.io/kubernetes/pkg/registry/flowcontrol/flowschema/storage.(*StatusREST).Update(0xc0887ccd50, {0x553da98, 0xc0887ccc00}, {0xc0412d2b3d, 0xc086dd5bf0}, {0x5523328, 0xc0887cccf0}, 0xc0991a2748, 0x40ef94, 0x1, ...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/flowcontrol/flowschema/storage/storage.go:93 +0x52
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*patcher).patchResource.func2()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:665 +0xa7
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.(*patcher).patchResource.func3()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:671 +0x38
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher.finishRequest.func1()
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:117 +0x8f
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher.finishRequest
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/finisher/finisher.go:92 +0xe5
```
Integration job link: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107456/pull-kubernetes-integration/1484156493719670784
PR: https://github.com/kubernetes/kubernetes/pull/107456

I searched for this error in CI and this PR, looks like we only see this error in this PR so far. So I am hoping it has not introduced any flake. (maybe we need to let more time to pass to start seeing flakes)

### What did you expect to happen?

We should not see this error `FieldManager must be installed to run apply`

### How can we reproduce it (as minimally and precisely as possible)?

I saw it only in the PR mentioned.

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
This brings back https://github.com/kubernetes/kubernetes/pull/107456 to see if it introduces flakes/timeouts in the integration tests

- reproduce the integration test timeout that this PR introduces
- check the very large log file of the integration test and determine the root cause

#### Which issue(s) this PR fixes:
Fixes #107727

#### Does this PR introduce a user-facing change?
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
suggestions:
https://github.com/kubernetes/kubernetes/blob/aa49dffc7f24dc31dd333869be8e6e3cdfc00af9/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L602

we probably have to remove `utilfeature.DefaultFeatureGate.Enabled(features.ServerSideApply` from the conditional

https://github.com/kubernetes/kubernetes/blob/aa49dffc7f24dc31dd333869be8e6e3cdfc00af9/staging/src/k8s.io/apiserver/pkg/features/kube_features.go#L216
https://github.com/kubernetes/kubernetes/blob/aa49dffc7f24dc31dd333869be8e6e3cdfc00af9/staging/src/k8s.io/apiserver/pkg/features/kube_features.go#L198

probably set `LockToDefault:true`

we will need to reproduce the issue first, apply these changes and show that it fixes the issue.

**Comment by wojtek-t** [comment]
@tkashem - I don't think we should be trying to reproduce by running integration tests at this point.

IIRC, this PR was causing panics of kube-apiserver - can you please start with checking the logs that those are not happening (or maybe rather first confirm it was the case previously)?

**Comment by tkashem** [comment]
> Init container clonerefs not ready: (state: running) Init container initupload not ready: (state: waiting, reason: "PodInitializing", message: "") Init container place-entrypoint not ready: (state: waiting, reason: "PodInitializing", message: "")

/test pull-kubernetes-integration

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:484]: not directly related to this PR, maybe we can return an error here

**Comment by tkashem** [comment]
the panics don't appear in integration log any longer, see https://github.com/kubernetes/kubernetes/issues/107727#issuecomment-1246156747

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:462]: @liggitt - we're generally locking GA features to true and SSA is already GA, but not locked:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/features/kube_features.go#L234

What was the reason for it?

[I'm asking because if it was locked, then we wouldn't need this conditional logic here.]

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:462]: cc @apelisse - I'd be in favor of locking/dropping the conditional apply gate logic

**Comment by tkashem** [comment]
> I just noticed this PR referenced in [#112306 (comment)](https://github.com/kubernetes/kubernetes/pull/112306#discussion_r970257918)
> 
> can we resolve the test flake in reset_fields_test.go and drop flowschemas from resetFieldsSkippedResources _before_ re-enabling SSA in the APF controller?

@liggitt are you referring to adding the `patchStrategy` annotation in Condition slices in `FlowSchemaStatus` and `PriorityLevelConfigurationStatus`  of `v1beta3` so strategic merge patch does not replace the entire conditions slice? https://github.com/kubernetes/kubernetes/issues/107574

Please see - https://github.com/kubernetes/kubernetes/issues/104842#issuecomment-1013578552

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:447]: Opened https://github.com/kubernetes/kubernetes/pull/112748

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:447]: @tkashem - this is now merged
Once the dependent PR are merged you can update this PR and get rid of this condition - this if will now always be true.

**Comment by MikeSpreitzer** [comment]
@tkashem : the reasons for the hold on this are gone, right?

**Comment by tkashem** [comment]
> the reasons for the hold on this are gone, right?

@MikeSpreitzer I wanted to make sure that https://github.com/kubernetes/kubernetes/pull/112575 resolves  https://github.com/kubernetes/kubernetes/issues/104842 

It's been a week, and https://storage.googleapis.com/k8s-triage/index.html?date=2022-10-01&job=unit&test=TestApplyResetFields yields zero result, so I am confident the flake has been resolved

Also with https://github.com/kubernetes/kubernetes/pull/112748 merged, SSA is always enabled, so I removed the conditiona `updateOrApply` method.

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM
Thanks!

**Comment by tkashem** [comment]
/hold cancel

all dependencies have merged, and search in ci yields yields zero fake

**Comment by MikeSpreitzer** [comment]
And thanks all for persistence in getting this done.

**Pull request merged**
